clusterloader/framework/rc_utils.go

```diff
@@ -0,0 +1,130 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/test/e2e/framework"
+)
+
+// CreateRC will create a new RC if it does not exist, or it will update an existing RC with a new replica count if it does exist
+func CreateRC(f *framework.Framework, namespace, name, imageName string, label map[string]string, replicas int) error {
+	fieldSelector := fields.OneTermEqualSelector("metadata.name", name)
+	list, _ := f.ClientSet.Core().ReplicationControllers(namespace).List(metav1.ListOptions{FieldSelector: fieldSelector.String()})
+	switch len(list.Items) {
+	case 0:
+		newRC := newRC(name, int32(replicas), label, name, imageName)
+		rc, err := createNewRCWithRetries(f, name, namespace, newRC)
+		if err != nil {
+			return err
+		}
+		framework.Logf("NEW RC: %+v", rc)
+	default:
+		rc, err := framework.UpdateReplicationControllerWithRetries(f.ClientSet, namespace, name, func(update *v1.ReplicationController) {
+			x := int32(replicas)
+			update.Spec.Replicas = &x
+		})
+		if err != nil {
+			return err
+		}
+		framework.Logf("UPDATED RC: %+v", rc)
+	}
+
+	pods, err := framework.PodsCreatedByLabel(f.ClientSet, namespace, name, int32(replicas), labels.SelectorFromSet(labels.Set(label)))
+	if err != nil {
+		return err
+	}
+
+	waitForRCPodsRunning(f, pods, name, namespace, replicas)
+	return nil
+}
+
+// createNewRCWithRetries uses polling to retry RC creation
+func createNewRCWithRetries(f *framework.Framework, name, namespace string, rcObj *v1.ReplicationController) (*v1.ReplicationController, error) {
+	var rc *v1.ReplicationController
+	var updateErr error
+	pollErr := wait.PollImmediate(10*time.Millisecond, 1*time.Minute, func() (bool, error) {
+		var err error
+		if rc, err = f.ClientSet.Core().ReplicationControllers(namespace).Create(rcObj); err == nil {
+			framework.Logf("Creating replication controller %q", name)
+			return true, nil
+		}
+		updateErr = err
+		return false, nil
+	})
+	if pollErr == wait.ErrWaitTimeout {
+		pollErr = fmt.Errorf("Unable to create new RC %q: %v", name, updateErr)
+	}
+	return rc, pollErr
+}
+
+// waitForRCPodsRunning will verify that a list of pods will all be running
+func waitForRCPodsRunning(f *framework.Framework, podList *v1.PodList, podName, namespace string, replicas int) error {
+	framework.Logf("Ensuring all pods for ReplicationController %q are running", podName)
+	running := 0
+	for _, pod := range podList.Items {
```

Comment:
The other thing is that it shouldn't be "poll-based". I mean you shouldn't periodially list pods (in general, you shouldn't list pods). Instead of that you should watch for them.

We already have a simple utility for it - see this:
https://github.com/kubernetes/kubernetes/blob/master/test/utils/pod_store.go#L37